### PR TITLE
Updated Contracts

### LINKED ISSUE #218: Add DAppChain Contracts Bindings & Clients
Initialize by name or address --> create general client class which implements the `connect` and `deploy` methods

- ERC20 (ETH is part of this)
- ERC721
- ERC721x
- Address Mapper
- DPoS
- Gateways
- Plasma Cash

### PR DESCRIPTION
**Pull request by gakonst**
Will close #218 

Changes are mostly refactoring and moving files from TG-v2 and Loom-Timelock to go-loom.

Features:
- [x] MirroredTokenContract is a wrapper that includes functionality for tokens 1:1 mapped between mainnet and dappchain.
- [x] ERC20
- [x] ERC721
- [x] ERC721X
- [x] NativeCoin (ETH or LOOM)
- [x] EVM_ETH
- [x] Gateway: https://github.com/loomnetwork/transfer-gateway-v2/pull/73
- [x] Timelock -> repo got much cleaner: https://github.com/loomnetwork/loom-timelock/commit/fb32d6b6d80afafc64a87d584e0a176f3e205079
- [x] DPOS

Will be added on a separate PR:
- [ ] Plasma Cash

### CODE REVIEW

File: client/erc20/dappchain_erc20_contract.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package erc20
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/erc20/dappchain_erc20_contract.go]: This file needs an `evm` build tag.

File: client/erc20/mainnet_erc20_client.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package erc20
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/erc20/mainnet_erc20_client.go]: `evm` build tag

File: client/erc721/dappchain_erc721_contract.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package erc721
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/erc721/dappchain_erc721_contract.go]: `evm` build tag

File: client/erc721/mainnet_erc721_contract.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
+package erc721
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/erc721/mainnet_erc721_contract.go]: `evm` build tag

File: client/erc721x/dappchain_erc721x_contract.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package erc721x
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/erc721x/dappchain_erc721x_contract.go]: `evm` build tag

File: client/erc721x/mainnet_erc721x_contract.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package erc721x
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/erc721x/mainnet_erc721x_contract.go]: `evm` build tag

File: client/evm_eth/dappchain_eth_evm_test_contract.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
+package evm_eth
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/evm_eth/dappchain_eth_evm_test_contract.go]: `evm` build tag

File: client/gateway/dappchain_gateway.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/gateway/dappchain_gateway.go]: `evm` build tag

File: client/gateway/mainnet_gateway_client.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import (
+	"context"
+	"crypto/ecdsa"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/gateway/mainnet_gateway_client.go]: `evm` build tag

File: client/helpers.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/helpers.go]: `evm` build tag

File: client/identity.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package client
+
+import (
+	"crypto/ecdsa"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
```
Comment (enlight) [inline client/identity.go]: `evm` build tag

File: client/mirrored_token_contract.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
+package client
+
+import (
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
```
Comment (enlight) [inline client/mirrored_token_contract.go]: `evm` build tag

File: client/timelock/mainnet_loom_timelock.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package timelock
+
+import (
+	"context"
+	"time"
+
+    "github.com/loomnetwork/go-loom/client"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/timelock/mainnet_loom_timelock.go]: `evm` build tag

Comment (enlight) [reply client/timelock/mainnet_loom_timelock.go]: And something is wrong with the whitespace

File: client/timelock/mainnet_timelock_factory_client.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
+package timelock
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
```
Comment (enlight) [inline client/timelock/mainnet_timelock_factory_client.go]: `evm` build tag & whitespace

File: Makefile [hunk 15]
```diff
@@ -82,6 +82,8 @@ deps:
 		golang.org/x/crypto/sha3 \
 		github.com/gogo/protobuf/jsonpb \
 		github.com/gogo/protobuf/proto \
+		github.com/gorilla/websocket \
+		github.com/phonkee/go-pubsub \
```
[the remark this answers is not in the document]
Comment (gakonst) [reply Makefile]: They are new deps required for the Gateway client's pubsub stuff

[the remark this answers is not in the document]
Comment (enlight) [reply Makefile]: The question is are they also used by go-loom? If not, they should be vendored instead of `go get`ed.

**Pull request merged**